feat: add oops.WrapfWithMetadata

danielcondemarin opened the pull request:
This feature allows decorating oops errors with arbitrary metadata which users can retrieve later for purposes such as logging.

- Metadata can be attached to oops errors. (`oops.WrapfWithMetadata`)
- Calling `oops.CollectMetadata` retrieves the aggregated metadata for the entire chain.

danielcondemarin committed:
feat: add oops.WrapfWithMetadata

This feature allows decorating oops errors with arbitrary metadata which users can retrieve later for purposes such as logging.

- Metadata can be attached to oops errors. (`oops.WrapfWithMetadata`)
- Calling `oops.CollectMetadata` retrieves the aggregated metadata for the entire chain.

jsm commented on oops/oops_test.go:
```diff
@@ -898,6 +898,67 @@ subsequent stack frames truncated
+			},
+		},
+		{
+			description: "metadata is aggregated from 2 oops errors and key deduplicated by selecting inner most value",
```
I think it would make more sense to use the outer-most value? WDYT?

danielcondemarin replied:
I thought of returning the inner most because that's the root cause of the error but I'm not too opinionated about that tbh. I think the better option might be to include everything and rename the key when duplicates are found but didn't wanna add the additional complexity unless is something we really think valuable. 
For now I'm happy to select the outer most: https://github.com/samsarahq/go/pull/32/commits/d8b91cfab71ba05798f3996252d97f811de9f71a

danielcondemarin committed:
feat: oops.CollectMetadata returns outer most key when dup

If CollectMetadata finds a duplicate key in the metadata, it returns the outer most key instead of inner most.

jsm commented on oops/oops.go:
```diff
@@ -72,6 +75,26 @@ func (e *oopsError) Error() string {
 	return b.String()
 }
 
+// CollectMetadata finds the first oopsError in err's chain and collects all metadata from oops errors in the chain.
```
Please add that "If multiple oops errors in the chain set a metadata field, the outer-most one's value is used"

danielcondemarin committed:
docs: improve oops methods documentation

Improves documentation for CollectMetadata and WrapfWithMetadata
